### PR TITLE
[ONNX] Introduce 'Functionalization' for fx exporter

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98246
* __->__ #98245

### CODE REVIEW

File: test/onnx/test_fx_dynamic_with_onnxruntime.py [hunk 1]
```diff
@@ -196,6 +196,18 @@ def forward(self, x, y):
             DynamicAdd(), (x, y), additional_test_inputs=[(input_x, input_y)]
         )
 
+    def test_mutation(self):
```
Comment (BowenBao) [inline test/onnx/test_fx_dynamic_with_onnxruntime.py:199]: To be deduplicated once dynamic tests are parameterized.

File: torch/onnx/_internal/fx/fx_exporter.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 from typing import Any, Callable, Mapping, Sequence
 
 import torch._ops
+import torch.fx
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/fx_exporter.py:5]: for intellisense/pylance not complaining `"fx" is not a known member of module "torch"` for `torch.fx.GraphModule`.

File: torch/onnx/_internal/fx/passes/utils.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/utils.py:1]: Should we limit the scope of this utils module by adding explicit guidance here?

**Review by justinchuby — COMMENTED** [review]
Let's go!!

File: torch/onnx/_internal/fx/passes/utils.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/utils.py:1]: wdym? could you elaborate

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    This pass utilizes ``torch.func.functionalize`` to convert a GraphModule into a
+    functional form. The two main functionalities are:
+
+    * ``torch.func.functionalize`` removes (intermediate) mutations and aliasing from a
+    function, while preserving the function's semantics.
+
+    * ``torch.func.functionalize`` also removes mutations (and views) that were performed
+    on function inputs. However to preserve semantics, functionalize will "fix up" the
+    mutations after the transform has finished running, by detecting if any tensor inputs
+    "should have" been mutated, and copying the new data back to the inputs if necessary.
```
Comment (wschin) [inline torch/onnx/_internal/fx/passes/functionalization.py:25]: How does `copying the new data back to the inputs` mean?

File: torch/onnx/_internal/fx/passes/utils.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/utils.py:1]: Right it was about the module. Sorry for the confusion

File: torch/onnx/_internal/fx/passes/utils.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        node for node in reference_module.graph.nodes if node.op == "placeholder"
+    ]
+
+    if len(placeholders) != len(reference_placeholders):
+        raise RuntimeError(
+            "The two modules have different number of arguments. "
+            f"module: {len(placeholders)}, reference_module: {len(reference_placeholders)}"
+        )
+
+    for placeholder, reference_placeholder in zip(placeholders, reference_placeholders):
```
Comment (wschin) [inline torch/onnx/_internal/fx/passes/utils.py:49]: Let's handle potential name conflicts here.
```
def foo(x1, y1):
  x=x1+y1
  b=a*x1 
```
and
```
def foo_reference(x, y):
  a=x+y
  b=a*x 
```

Either
1. Assert if any names from `reference_module` are being used.
2. Change the name `x` used in `foo` to `x_another` and then you can replace `x1` with `x`.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:25]: Consider
```python
def fn(a, b):
    a.add_(b)
    return a
```
When you call this function `fn(x, y)`, `x` is also mutated. So just functionalizing is not enough for preserving the original semantics. A "special" input mutation step, if needed, needs to be inserted at the end. That's what the comment means. 
```python
# after functionalization
# this is not semantically the same. `a`'s value is not updated.
def fn(a, b):
    new_a = a + b
    return new_a

# functionalization that preserves semantics
def fn(a, b):
    new_a = a + b

    # copying the new data back to the inputs
    a.copy_(new_a)

    return new_a
```

File: torch/onnx/_internal/fx/passes/utils.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/utils.py:49]: Good catch, this function is unsafe. I was just moving this function in this PR. For the sake of functionalization scoping, I'm happy to expand on this in another PR.

File: torch/onnx/_internal/fx/passes/utils.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/utils.py:1]: In addition to docstring, what would be a better file name, `utils.py`, `common.py`, `common_utils.py`, etc?

Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/utils.py:1]: I think we can import passes and use `passes.utils`. `passes.common` sounds great too

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/utils.py:1]: I'll go with `passes._utils` then, they are not supposed to be used outside lol.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:25]: added above into docstring

File: torch/onnx/_internal/fx/passes/utils.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/utils.py]: added above into docstring

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    This pass is recommended to be used after ``Functionalization`` pass.
+    ``Functionalization`` pass adds `aten.copy_.default` nodes to the graph
+    when it detects mutations to inputs. These nodes are not needed for ONNX export
+    for inference. They could be useful for training.
+    """
+
+    @_beartype.beartype
+    def _run(self, *args) -> torch.fx.GraphModule:
+        for node in reversed(self.module.graph.nodes):
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/passes/functionalization.py:95]: Why do we need reversed iterator?

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            new_a = a + b
+
+            # Copying the new data back to the inputs
+            a.copy_(new_a)
+
+            return new_a
+
+    For ONNX inference, it is recommended to run ``RemoveInputMutation`` after this pass.
+    ``RemoveInputMutation`` removes the "fix up" nodes that were added by ``Functionalize``,
+    which are not needed for ONNX inference.
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/passes/functionalization.py:54]: Why not remove fix up in this pass?

File: torch/onnx/_internal/fx/fx_exporter.py [hunk 8]
```diff
@@ -24,21 +25,27 @@ def export_fx_to_onnx(
...
     ) -> torch.onnx.ExportOutput:
         # Apply decomposition table to the input graph.
-        decomposed_module = passes.Decompose(
+        module = passes.Decompose(
             fx_module,
             self.decomposition_table,
             enable_dynamic_axes=self.options.dynamic_shapes,
         ).run(*fx_module_args)
 
+        # Resolve mutations.
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/fx_exporter.py:34]: Can we have an explanation why we need to resolve mutations somewhere in the codebase? It's for future maintenance.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:95]: Here one of the conditions to meet when deleting a node is it is unused by others. The graph is topologically sorted, meaning user node always comes after used node. So deleting from backwards ensures to delete the nodes whose users are all deleted.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:54]: Because input mutation might be useful in training.

File: torch/onnx/_internal/fx/fx_exporter.py (hunk 8 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/fx_exporter.py:34]: There is a test case added. It fails otherwise.

Comment (BowenBao) [reply torch/onnx/_internal/fx/fx_exporter.py:34]: I misunderstood your comment... Sure.

Comment (BowenBao) [reply torch/onnx/_internal/fx/fx_exporter.py:34]:
```
        # ONNX does not support views and mutations.
        # Functionalize to get a semantically equivalent graph without mutations.
```
wdyt?

**Review by wschin — APPROVED** [review]

File: torch/onnx/_internal/fx/passes/utils.py (hunk 5 above)
Comment (wschin) [reply torch/onnx/_internal/fx/passes/utils.py]: Another PR is fine. Just want to note that at large models such as GPT, name conflict can easily occur. :)

File: torch/onnx/_internal/fx/fx_exporter.py (hunk 8 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/fx_exporter.py:34]: Thanks!

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 7 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/functionalization.py:54]: So we are likely to bypass the RemoveInputMutation pass if model is in training mode?

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:54]: Yes.

To add more details, I think how to make use of this information depends on how `dynamo_export` integrates with training. But generally speaking, it is always better to preserve as much information as possible.

**Review by titaiwangms — APPROVED** [review]
LGTM!
